---
title: "NSBB 552 Final Project"
output: html_notebook
---

## Background:

##### Necrotizing enterocolitis (NEC) is a gastrointestinal disease that is the leading cause of premature infant death in the NICU, with a mortality rate of 20--50%. NEC is difficult to diagnose due to late stage signs such as abdominal distension, blood in stool, microbial imbalance, and gas in intestinal walls (pneumatosis intestinalis) which quickly progresses to local and systemic inflammation, multi-organ failure, and death. Even if the premature infant survives NEC, they may lead a decreased quality of life due to permanent bowel issues and neurodevelopmental delays from NEC exposure. The NEC profile is characterized by up-regulation of pro-inflammatory markers such as IL-6, IL-8, IL-1B, and TNF alpha.

## Hypothesis:

##### We hypothesize that the development and severity of NEC can be attenuated by through inhibition of pro-inflammation pathways.

### Dataset Information

##### link: <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE154617>

##### SRA: SRP272342

##### file type: fastq.gz

##### Summary: Comparison of intestinal epithelial cell tissue of breast fed (control) and formula fed (NEC model) mice at P4

##### Importing Libraries

```{r}
library(tximport)
library(readr)
library(tximportData)
library(DESeq2) #nra-seq analysis 
library(pheatmap)
library(vsn)
library(RColorBrewer)
library(gplots)
library(ggplot2)
library(genefilter)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(dplyr)
library(DiagrammeR)
library(GenomicFeatures)
library(apeglm)
library(gage)
library(gageData)
library(pathview)
library(biomaRt)
library(clusterProfiler)
```

### Design

```{r}
grViz("digraph flowchart {
      # node definitions with substituted label text
      node [fontname = Helvetica, shape = rectangle]        
      tab1 [label = '@@1']
      tab2 [label = '@@2']
      tab3 [label = '@@3']
      tab4 [label = '@@4']
      tab5 [label = '@@5']

      # edge definitions with the node IDs
      tab1 -> tab2 -> tab3 -> tab4 -> tab5;
      }

      [1]: 'Obtain data'
      [2]: 'Process and clean raw data: fastq to readCounts'
      [3]: 'Observe sample quality'
      [4]: 'Data Visualization'
      [5]: 'Pathway Analysis' 
      ")
```

# **Obtain data**

### Downloading raw, zipped files from NCBI GEO repository

```{bash}
#process ()
#{
#        fasterq-dump $SRR
#        mv ${SRR}_1.fastq ${sample}_1.fastq; mv ${SRR}_2.fastq ${sample}_2.fastq
#        gzip ${sample}_1.fastq  ${sample}_2.fastq
#}

#for GSM in $(grep -i -v i3c GSMtable.txt | cut -f1)
#do
#        SRR=$(grep $GSM runInfo.txt | cut -f1)
#        sample=$(grep $GSM GSMtable.txt | cut -f2 | awk '{gsub(" ", ""); print}')
#        #echo $sample
#        process &
#done
#wait
#echo done 

```

### Importing raw files into alignment tool (Salmon)

```{bash}
#for fn in sample.e144_14_ff sample.e144_17_ff sample.e144_19_ff sample.e144_1_bf sam>
#do  
#echo "Processing sample $fn"
#salmon quant -i /home/aaltamirano/Documents/nsbb552/genome_folder/alias/mm10/salmon_>
#        -1 ${fn}_1.fastq \
#        -2 ${fn}_2.fastq \
#        -p 24 --validateMappings -o quants/Salmon/${fn}
#done 
#echo "done" 
```

### Setting up files for DESeq2 Analysis

```{r}
# Setting working directory
dir <- "/home/aaltamirano/Documents/nsbb552/quants/Salmon"
list.files(dir)
```

```{r}
# Make a gene ID x Transcript name data frame from reference genome (tx2gene)
# Order of columns matters 
txdb <- makeTxDbFromGFF("/home/aaltamirano/Documents/nsbb552/quants/Salmon/0f10d83b1050c08dd53189986f60970b92a315aa7a16a6f1.gtf")
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- AnnotationDbi::select(txdb, keys = k, keytype = "TXNAME", columns = "GENEID")
head(tx2gene)
```

```{r}
#Make a samples.txt file with sample IDs - should match the sample IDs used for alignment/mapping
#Assign samples.txt to the samples variable
samples <- read.table(file.path(dir, "samples.txt"), header = TRUE)
samples$condition <- factor(rep(c("A","B"),each=4)) #Adds a condition column
samples
# A=control B=NEC


files <- file.path(dir, samples$sample, "quant.sf")
names(files) <- paste0(samples$sample)
#files
txi.salmon <- tximport(files, type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE)
#names(txi.salmon)
head(txi.salmon$counts)

```

```{r}
#all(file.exists(files))
#head(files)
#file.exists(files)
data.frame(thefiles = files, doihave = file.exists(files))
```

### Loading deseq2

```{r}
dds <- DESeqDataSetFromTximport(txi.salmon, samples, ~condition)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```

```{r}
dds$condition <- factor(dds$condition, levels = c("A","B"))
```

# **Processing data**

```{r}
dds <- DESeq(dds)
dds
res <- results(dds)
#res <- res[res$log2FoldChange >= 1, ]
summary(res)
```

# **Clean data**

```{r}
## log fold change shrinkage for visualization and ranking
resultsNames(dds)
resLFC <- lfcShrink(dds, coef="condition_B_vs_A", type="apeglm")
resLFC
```

```{r}
## Organizing results by smallest P-value
resOrdered <- res[order(res$pvalue),]
resOrdered

## How many adjusted p-values were less than 0.01?
sum(res$padj < 0.1, na.rm=TRUE)

## Changing p-value to 0.05 
res05 <- results(dds, alpha=0.05)
summary(res05)
## How many adjusted p-values were less htan 0.05?
sum(res05$padj < 0.05, na.rm=TRUE)
```

# **Observe quality of data**

```{r}
plotMA(res, ylim=c(-2,2))
plotMA(resLFC, ylim=c(-2,2))
```

```{r}
#checked 
plotDispEsts(dds, main="Dispersion plot")
```

```{r}
mcols(resLFC)$description
```

```{r}
ntd <- normTransform(dds)
rld <- rlogTransformation(dds)
head(assay(rld))

vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)

meanSdPlot(assay(ntd))
meanSdPlot(assay(vsd))
meanSdPlot(assay(rld))

#Shows the effect of the transformation, i
ddsESF <- estimateSizeFactors(dds)
df1 <- data.frame(log2(counts(ddsESF, normalized=TRUE)[, 1:2] + 1))
df1$transformation <- "log2(x + 1)"
df2 <- data.frame(assay(rld)[, 1:2])
df2$transformation <- "rld"
df3 <- data.frame(assay(vsd)[, 1:2])
df3$transformation <- "vsd"
df <- rbind(df1, df2, df3)
colnames(df)[1:2] <- c("x", "y")
head(df)
table(df$transformation)

options(repr.plot.width=6, repr.plot.height=2.5)
ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
coord_fixed() + facet_grid( . ~ transformation)
#rld graph compresses differences for the low count genes > excludes genes with low counts

par(mfrow=c(1,3))
boxplot(log2(assay(ddsESF)+1), las=2, main="log2(x+1)")
boxplot(assay(rld), las=2, main="rld")
boxplot(assay(vsd), las=2, main="vsd")
```

```{r}

select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)["condition"])
pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
```

```{r}
#
(mycols <- brewer.pal(8, "Dark2")[1:length(unique(samples$condition))])
sampleDists <- as.matrix(dist(t(assay(rld))))
heatmap.2(as.matrix(sampleDists), key=F, trace="none",
          col=colorpanel(100, "blue", "white"),
          ColSideColors=mycols[samples$condition],
          RowSideColors=mycols[samples$condition],
          margin=c(10, 10), main="Sample Distance Matrix")
```

```{r}
sampleDists <- dist(t(assay(rld)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition, rld$sample, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

```{r}
#PCA Alternative code
#DESeq2::plotPCA(rld, intgroup="condition")
```

```{r}
pcaData <- plotPCA(rld, intgroup="condition", returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=condition, shape=condition)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

```{r}
table(resLFC$padj<0.05)
res <- resLFC[order(resLFC$padj), ]
resdata <- merge(as.data.frame(res), as.data.frame(counts(dds, normalized=TRUE)), by="row.names", sort=FALSE)
names(resdata)[1] <- "Gene"
head(resdata)
```

# **Data Visualization**

```{r}
# Volcano plot
with(resLFC, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot",ylim=c(0,10), xlim=c(-2,2)))
# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(res, padj<.05 ), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
with(subset(res, abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="orange"))
with(subset(res, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="green"))
```

```{r}
#Assigning the top 20 significant genes 
topVargenes20 <- head(order(rowVars(assay(rld)), decreasing = TRUE), 20)
topVargenes500 <- head(order(rowVars(assay(rld)), decreasing = TRUE), 500)


#Making a heatmap of the top 20 significant genes 
mat <- assay(rld)[topVargenes20, ]
mat <- mat - rowMeans(mat)
anno <- as.data.frame(colData(rld)["condition"])
pheatmap(mat, annotation_col = anno)


mat <- assay(rld)[topVargenes500, ]
mat <- mat - rowMeans(mat)
anno <- as.data.frame(colData(rld)["condition"])
pheatmap(mat, scale = "row", show_rownames = F,
         clustering_method = 'average',  annotation_col = anno, cutree_cols = 3)
```

# **Pathway Analysis**

```{r}
#Creating datasets with KEGG gene sets to test
kegg_mouse <- kegg.gsets(species = "mouse", id.type = "kegg")
names(kegg_mouse)
kegg.gs <- kegg_mouse$kg.sets[kegg_mouse$sigmet.idx]
```

```{r}
res$symbol = mapIds(org.Mm.eg.db,
                     keys=row.names(res), 
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
res$entrez = mapIds(org.Mm.eg.db,
                     keys=row.names(res), 
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")
res$name =   mapIds(org.Mm.eg.db,
                     keys=row.names(res), 
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
```

```{r}
#Running GAGE

keggres = gage(foldchanges, gsets=kegg.gs, same.dir = T)
names(keggres)

lapply(keggres, head) #Up-regulated and down-regulated pathways
#head(keggres$greater) #Up-regulated pathways
#head(keggres$less) # Down-regulated pathways

# Explore the top 20 up-regulated pathways and KEGG IDs
keggrespathways = data.frame(id=rownames(keggres$greater), keggres$greater) %>% 
  tbl_df() %>% 
  filter(row_number()<=20) %>% 
  .$id %>% 
  as.character()
keggrespathways
# Get the IDs.
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

```{r}
# Define plotting function for applying later
plot_pathway = function(pid) pathview(gene.data = foldchanges, pathway.id=pid, species="mouse", new.signature=FALSE)

#plot multiple pathways (plots saved to disk and returns a throwaway list object)
tmp = sapply(keggresids, function(pid) pathview(gene.data = foldchanges, pathway.id=pid, species="mouse"))
```

```{bash}
# Open inflammation-associated pathway .png files
feh ~/Documents/nsbb552/mmu04662.pathview.png #"mmu04662 B cell receptor signaling pathway"
feh ~/Documents/nsbb552/mmu04657.pathview.png #"mmu04657 IL-17 signaling pathway"
feh ~/Documents/nsbb552/mmu04672.pathview.png #"mmu04672 Intestinal immune network for IgA production"
feh ~/Documents/nsbb552/mmu04668.pathview.png #"mmu04668 TNF signaling pathway"  

```

# **Discussion**

##### After processing and cleaning the dataset, there is some clear differential expression between the breast fed (Condition A: Control) and the formula fed (Condition B: NEC) groups. This difference is most clearly observed when comparing samples 2,3,4 and 14,15,17, and 19. The top 20 most differentiated genes didn't have a strong association with inflammation pathways. Aditionally, downstream pathway analyses of p vale\<0.05 were to stringent, so p-value\<0.1 was used. In conlcusion, this dataset does show differntial expression of genes in pro-inflammation pathways. Further pathways analyses with IPA and possible gene enrichment would be useful in pursuing these connections.

```{r}
## Doesn't make biological sense 
# Create background dataset for hypergeometric testing using all genes tested for significance in the results                  
all_genes <- as.character(rownames(res)) 
# Extract significant results 
signif_res <- res[res$padj < 0.07 & !is.na(res$padj), ] 
signif_genes <- as.character(rownames(signif_res)) 

ego <- enrichGO(gene = signif_genes, universe = all_genes, keyType = "ENSEMBL", OrgDb = org.Mm.eg.db, ont = "BP",pAdjustMethod = "BH", qvalueCutoff = 0.07, readable = TRUE) 
# Output results from GO analysis to a table 
cluster_summary <- data.frame(ego) 
#Visualizing 
dotplot(ego, showCategory=50)
#emapplot(ego, showCategory=50)

# To color genes by log2 fold changes 
signif_res_lFC <- signif_res$log2FoldChange 
cnetplot(ego, categorySize="pvalue", showCategory = 5, foldChange= signif_res_lFC, vertex.label.font=6)
```

```{r}
sessionInfo()
```

### Relevant Documentation

1.  rnaseqGene: end-to-end gene-level RNA-seq differential expression workflow using Bioconductor packages (<http://master.bioconductor.org/packages/release/workflows/html/rnaseqGene.html>)

```{r}
browseVignettes("rnaseqGene")
```

2.  DESeq2: Differential gene expression analysis based on the negative binomial distribution (<https://bioconductor.org/packages/release/bioc/html/DESeq2.html>)

```{r}
browseVignettes("DESeq2")
```

3.  Salmon: A tool for quantifying the expression of transcripts using RNA-seq data (<https://combine-lab.github.io/salmon/getting_started/#after-quantification>)

-   

    ###### Patro, R., Duggal, G., Love, M. I., Irizarry, R. A., & Kingsford, C. (2017). Salmon provides fast and bias-aware quantification of transcript expression. Nature Methods.

4.  apeglm

-   

    ###### Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for

    sequence count data: removing the noise and preserving large differences. Bioinformatics. <https://doi.org/10.1093/bioinformatics/bty895>
